Update material-ui_v1.x.x.js

wcandillon opened the pull request:
This is PR improve the typing for `withStyle` and introduces the `Theme` type.

wcandillon commented:
@bugzpodder Could you take a look? Looking forward to read your thoughts on this. I'm hope we can make flow support in material-ui as solid as its TS support in the long run 💪🏻

bugzpodder commented:
Please fix the CI build, I'll take a look at the change today.

wcandillon commented:
@bugzpodder Fixed the build, thks a lot

wcandillon commented on definitions/npm/material-ui_v1.x.x/flow_v0.58.x-/material-ui_v1.x.x.js:
```diff
@@ -1,3 +1,5 @@
+import * as React from "react";
```
Yes that would makes sense to remove it if `React$ElementConfig` is available under that syntax. I will check.

wcandillon replied:
According to the thread below, this seems to be the recommended approach: https://github.com/flowtype/flow-typed/issues/1302

wcandillon commented on definitions/npm/material-ui_v1.x.x/flow_v0.58.x-/material-ui_v1.x.x.js:
```diff
@@ -1611,28 +1629,148 @@ declare module "material-ui/styles/createGenerateClassName" {
 declare module "material-ui/styles/createMixins" {
+
+  declare export type Mixins = {
+      gutters: (styles?: React.CSSProperties) => React.CSSProperties,
```
good point. Will fix it.

wcandillon replied:
@bugzpodder I would like this type file to depend on https://github.com/frenic/csstype, how do we handle such dependencies in flowtype?

wcandillon commented on definitions/npm/material-ui_v1.x.x/flow_v0.58.x-/material-ui_v1.x.x.js:
```diff
@@ -1611,28 +1629,148 @@ declare module "material-ui/styles/createGenerateClassName" {
+  declare type TypographyStyle = {
+    color?: string,
+    fontFamily: string,
+    fontSize: string,
```
yes 👍🏻. I took it from the typing of CSSProperties in flow but it can be `string` | `number`, will fix it.

wcandillon commented on definitions/npm/material-ui_v1.x.x/flow_v0.58.x-/material-ui_v1.x.x.js:
```diff
@@ -1611,28 +1629,148 @@ declare module "material-ui/styles/createGenerateClassName" {
+      shadows: Shadows,
+      transitions: Transitions,
+      spacing: Spacing,
+      zIndex: ZIndex
```
Yes it's a big one to add: https://github.com/mui-org/material-ui/blob/v1-beta/packages/material-ui/src/styles/overrides.d.ts
But I can try to do it.

bugzpodder replied:
I think we can just do overrides?: Object?  Thanks for the hard work and no need to cram everything in this change :)

wcandillon replied:
haha yes absolutely. Really appreciate your support 👍

wcandillon commented on definitions/npm/material-ui_v1.x.x/flow_v0.58.x-/material-ui_v1.x.x.js:
```diff
@@ -1644,6 +1782,15 @@ declare module "material-ui/styles/getStylesCreator" {
+  declare type Theme = MuiTheme;
+
+  declare type WithStyles<ClassKey: string> = {
+      classes: { [className: ClassKey]: string }
```
yes good point 👍🏻

wcandillon commented on definitions/npm/material-ui_v1.x.x/flow_v0.58.x-/material-ui_v1.x.x.js:
```diff
@@ -1658,11 +1805,45 @@ declare module "material-ui/styles/MuiThemeProvider" {
 
 declare module "material-ui/styles/shadows" {
+
+  declare export type Shadows = [
```
I agree 👍🏻

wcandillon commented on definitions/npm/material-ui_v1.x.x/flow_v0.58.x-/material-ui_v1.x.x.js:
```diff
@@ -1671,41 +1852,60 @@ declare module "material-ui/styles/themeListener" {
+  declare module.exports: <ClassKey: string, Props: {}, Comp: React.ComponentType<Props>>(
+      style: StyleRules<ClassKey> | StyleRulesCallback<ClassKey>,
+      options?: WithStylesOptions
+  ) => (C: Comp) => $Diff<React.ElementConfig<Comp>, WithStyles<ClassKey>>;
```
It works great on my project. Since `React.ElementConfig` has been released, expressing higher order components has been great 🙌🏻

wcandillon commented on definitions/npm/material-ui_v1.x.x/flow_v0.58.x-/material-ui_v1.x.x.js:
```diff
@@ -1671,41 +1852,60 @@ declare module "material-ui/styles/themeListener" {
+
+  declare type StyleRules<ClassKey: string = string> = { [className: ClassKey]: React.CSSProperties };
+  declare type StyleRulesCallback<ClassKey: string = string> = (theme: Theme) => StyleRules<ClassKey>;
+  declare type WithStyles<ClassKey: string> = {
```
yes thanks 👍🏻

AndrewSouthpaw commented:
Once the review process is finished, message me again. :)

wcandillon commented:
@bugzpodder Thank you for your amazing review, very thorough and I really like where we are going with this project 👍🏻. I addressed all review comments. Added an extra dependency with `csstype` which is automatically available if you have material-ui installed. Not sure how to deal with such dependencies in flow-types 🤔

Looking forward to read your thoughts on this.

wcandillon commented:
@bugzpodder I have two questions at this point:
* How should we deal with the `csstype` dependency? Should we copy the `csstype` type definition into the folder?
* What's the equivalent of `Record` in flow?  I wrote:
`  declare type StyleRules<ClassKey: string> = { [className: ClassKey]: $Exact<CSSProperties<number>> };` which is not exactly correct since `StyleRules` should have *all* the keys from `ClassKey` present.

Thank you for your support.

bugzpodder commented:
> How should we deal with the csstype dependency? Should we copy the csstype type definition into the folder?
Looks like there is a lot of stuff, not sure how easy it is to import it: https://github.com/frenic/csstype/blob/master/index.js.flow
We also have this, it may help referencing things like $PropertyType<CSSProperties<>, "fontSize">,
https://github.com/facebook/flow/tree/v0.70.0/lib/cssom.js#L71


> What's the equivalent of Record in flow
https://github.com/niieani/typescript-vs-flowtype#records

wcandillon commented:
@bugzpodder I looked at https://github.com/niieani/typescript-vs-flowtype#records but it doesn't seen to work as expected. Do you have an example that works that you could share via https://flow.org/try/?

bugzpodder commented:
I tried a bunch for Record, couldn't get it to work either.  Like you said $Exact doesn't really do what it is supposed to.  Sorry I just assumed it would work and never tried it out :(
